app/src/main/java/com/geeksville/mesh/MeshServiceClient.kt

```
@@ -0,0 +1,114 @@
+     * LifecycleOwner which overrides the default binding for @ActivityScoped. The solution to this is to add a
+     * qualifier to the LifecycleOwner provider in ApplicationModule.
+     */
+    private val activity: Activity,
+    private val serviceRepository: ServiceRepository,
+) : ServiceClient<IMeshService>(IMeshService.Stub::asInterface),
+    DefaultLifecycleObserver {
+
+    // TODO Use the default binding for @ActivityScoped
+    private val lifecycleOwner: LifecycleOwner = activity as LifecycleOwner
+
+    // TODO Inject this for ease of testing
+    private var serviceSetupJob: Job? = null
+
+    init {
+        debug("Adding self as LifecycleObserver for $lifecycleOwner")
+        lifecycleOwner.lifecycle.addObserver(this)
+    }
+
+    // region ServiceClient overrides
+
+    override fun onConnected(service: IMeshService) {
+        serviceSetupJob?.cancel()
+        serviceSetupJob =
+            lifecycleOwner.lifecycleScope.handledLaunch {
+                serviceRepository.setMeshService(service)
+                debug("connected to mesh service, connectionState=${serviceRepository.connectionState.value}")
+            }
+    }
+
+    override fun onDisconnected() {
+        serviceSetupJob?.cancel()
+        serviceRepository.setMeshService(null)
+    }
+
+    // endregion
+
+    // region DefaultLifecycleObserver overrides
+
+    override fun onStart(owner: LifecycleOwner) {
+        super.onStart(owner)
+        debug("Lifecycle: ON_START")
+
+        try {
+            bindMeshService()
+        } catch (ex: BindFailedException) {
+            errormsg("Bind of MeshService failed: ${ex.message}")
+        }
+    }
+
+    override fun onDestroy(owner: LifecycleOwner) {
+        super.onDestroy(owner)
+        debug("Lifecycle: ON_DESTROY")
+
+        owner.lifecycle.removeObserver(this)
+        debug("Removed self as LifecycleObserver to $lifecycleOwner")
+    }
+
+    // endregion
+
+    @Suppress("TooGenericExceptionCaught")
```

This is how the exception was caught before, so suppressing to pass detekt

---

is there a more specific exception type to handle here?

---

I'm not sure. `.startService()` appears to branch on API level only to call the same method in both cases. It swallows an exception for only one of them 🤷

---

My intention was to leave the underlying logic alone., and just refactor it.